File: packages/jetbrains-plugin/src/main/java/com/chainlesschain/ide/ApprovalSettlementRegistry.java
package com.chainlesschain.ide;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Process-local settlement guard for approval transport cards.
 *
 * <p>The CLI remains the durable authority. This registry only prevents one
 * JetBrains card from writing duplicate or conflicting responses while the
 * authoritative result is in flight, and lets a failed transport write return
 * the card to a retryable state.
 */
public final class ApprovalSettlementRegistry {
    public enum Status {
        PENDING,
        RESPONDING,
        INTERRUPTING
    }

    private final Map<String, Status> approvals = new LinkedHashMap<>();

    /** Register a newly rendered transport card. Duplicate ids stay unchanged. */
    public synchronized boolean open(String id) {
        String normalized = normalize(id);
        if (normalized == null || approvals.containsKey(normalized)) return false;
        approvals.put(normalized, Status.PENDING);
        return true;
    }

    /** Atomically reserve one pending card for an approval or denial response. */
    public synchronized boolean beginDecision(String id) {
        return begin(id, Status.RESPONDING);
    }

    /** Atomically reserve every pending card before writing an interrupt. */
    public synchronized List<String> beginInterrupt() {
        List<String> reserved = new ArrayList<>();
        for (Map.Entry<String, Status> entry : approvals.entrySet()) {
            if (entry.getValue() == Status.PENDING) {
                entry.setValue(Status.INTERRUPTING);
                reserved.add(entry.getKey());
            }
        }
        return reserved;
    }

    /**
     * Complete a transport attempt only when its reservation is still current.
     * Accepted decisions stay reserved until the CLI emits approval_resolved;
     * accepted interrupts invalidate their transport cards immediately.
     */
    public synchronized boolean complete(String id, Status reservation, boolean accepted) {
        String normalized = normalize(id);
        if (normalized == null || reservation == null || reservation == Status.PENDING
                || approvals.get(normalized) != reservation) return false;
        if (!accepted) {
            approvals.put(normalized, Status.PENDING);
        } else if (reservation == Status.INTERRUPTING) {
            approvals.remove(normalized);
        }
        return true;
    }

    /** Apply the CLI's authoritative resolution and forget the transport card. */
    public synchronized boolean resolve(String id) {
        String normalized = normalize(id);
        return normalized != null && approvals.remove(normalized) != null;
    }

    /** Invalidate every card when its owning child/session is stopped or replaced. */
    public synchronized void invalidateAll() {
        approvals.clear();
    }

    public synchronized int size() {
        return approvals.size();
    }

    public synchronized Status status(String id) {
        String normalized = normalize(id);
        return normalized == null ? null : approvals.get(normalized);
    }

    private synchronized boolean begin(String id, Status reservation) {
        String normalized = normalize(id);
        if (normalized == null || approvals.get(normalized) != Status.PENDING) return false;
        approvals.put(normalized, reservation);
        return true;
    }

    private static String normalize(String id) {
        if (id == null) return null;
        String normalized = id.trim();
        return normalized.isEmpty() ? null : normalized;
    }
}

File: android-app/core-agent-protocol/src/main/java/com/chainlesschain/android/core/agentprotocol/ApprovalSettlementRegistry.kt
package com.chainlesschain.android.core.agentprotocol

/**
 * Process-local CAS guard for Remote Session approval cards.
 *
 * The CLI remains the durable HumanTask authority. This registry only prevents
 * one Android transport card from sending duplicate/conflicting responses and
 * rolls a failed transport reservation back to a retryable pending state.
 */
class ApprovalSettlementRegistry(
    private val maxResolvedIds: Int = 1_024,
) {
    enum class Status {
        PENDING,
        RESPONDING,
        INTERRUPTING,
    }

    private val approvals = linkedMapOf<String, Status>()
    private val resolvedIds = linkedSetOf<String>()

    @Synchronized
    fun open(requestId: String): Boolean {
        val id = normalize(requestId) ?: return false
        if (id in approvals || id in resolvedIds) return false
        approvals[id] = Status.PENDING
        return true
    }

    @Synchronized
    fun beginDecision(requestId: String): Boolean =
        begin(requestId, Status.RESPONDING)

    @Synchronized
    fun beginInterrupt(): List<String> = approvals.entries
        .filter { it.value == Status.PENDING }
        .onEach { it.setValue(Status.INTERRUPTING) }
        .map { it.key }

    @Synchronized
    fun complete(requestId: String, reservation: Status, accepted: Boolean): Boolean {
        val id = normalize(requestId) ?: return false
        if (reservation == Status.PENDING || approvals[id] != reservation) return false
        if (!accepted) {
            approvals[id] = Status.PENDING
        } else if (reservation == Status.INTERRUPTING) {
            approvals.remove(id)
            rememberResolved(id)
        }
        return true
    }

    @Synchronized
    fun resolve(requestId: String): Boolean {
        val id = normalize(requestId) ?: return false
        val removed = approvals.remove(id) != null
        rememberResolved(id)
        return removed
    }

    @Synchronized
    fun invalidateAll() {
        approvals.clear()
        resolvedIds.clear()
    }

    @Synchronized
    fun size(): Int = approvals.size

    @Synchronized
    fun status(requestId: String): Status? = normalize(requestId)?.let(approvals::get)

    @Synchronized
    private fun begin(requestId: String, reservation: Status): Boolean {
        val id = normalize(requestId) ?: return false
        if (approvals[id] != Status.PENDING) return false
        approvals[id] = reservation
        return true
    }

    private fun rememberResolved(requestId: String) {
        resolvedIds += requestId
        while (resolvedIds.size > maxResolvedIds.coerceAtLeast(1)) {
            resolvedIds.remove(resolvedIds.first())
        }
    }

    private fun normalize(requestId: String?): String? =
        requestId?.trim()?.takeIf(String::isNotEmpty)
}
